### ISSUE TITLE
Unable to set IdentitySource to empty value / null

### ISSUE THREAD

**Issue by samuelAle** [serverless/serverless#9458]
The AWS::ApiGateway::Authorizer translation contains a bug.

The property [IdentitySource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-authorizer.html#cfn-apigateway-authorizer-identitysource) is only required when caching is enabled.

However, the serverless framework is [setting a default value](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/package/compile/events/apiGateway/lib/validate.js#L282) whenever it is undefined in the `serverless.yml`

Instead the logic should be updated to the following:

```javascript
    if (typeof identitySource === 'undefined' && resultTtlInSeconds !== 0) {
      identitySource = 'method.request.header.Authorization';
    }
```

<details>
<summary><code>serverless.yml</code></summary>

```yaml
provider:
  name: aws
  runtime: nodejs14.x
  timeout: 10
  memorySize: 256
  role: LambdaExecutionRole

functions:
  authorizer:
    handler: src/auth.authorize
  authHandler:
    handler: src/auth.handle
    events:
      - http:
          path: auth
          method: get
          authorizer:
            name: authorizer
            type: request
            resultTtlInSeconds: 0 # should disable caching

resources:
  Resources:
    LambdaExecutionRole:
      Type: AWS::IAM::Role
      Properties:
        RoleName: foo-role
        Path: /
        ManagedPolicyArns:
          - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
        AssumeRolePolicyDocument:
          Version: '2012-10-17'
          Statement:
            - Effect: Allow
              Principal:
                Service:
                  - lambda.amazonaws.com
              Action: sts:AssumeRole
```

</details>

<details>
<summary><b><code>⚠️⚠️ REPLACE WITH FULL COMMAND NAME</code> output</b></summary>

```
serverless deploy
```

</details>

<b>Installed version</b>

```
Serverless: Running "serverless" installed locally (in service node_modules)
Framework Core: 1.83.3 (local)
Plugin: 3.8.4
SDK: 2.3.2
Components: 2.34.9
```

**Comment by pgrzesik** [comment]
Hello @samuelAle, thanks for reporting. I'm wondering if we should set the default value at all. In case of `httpApi`, we throw an error when the `identitySource` is not defined when caching is enabled (via `resultTtlInSeconds`), what is your opinion on that?

**Comment by pgrzesik** [comment]
> However, I also don't want existing projects to suddenly see exceptions after upgrading to the latest version of the serverless framework that contains this fix.

This is definitely not something that we would like to introduce. 

As for the default value itself, for `REQUEST` type when caching is not enabled, is `identitySource` setting affecting anything meaningful? As far as I understand, it is used when caching is involved to construct the cache-key.

**Comment by samuelAle** [comment]
Yes, even when caching is disabled, AWS API Gateway will reject any requests that do not contain the specified IdentitySource.

**Comment by pgrzesik** [comment]
Thanks for clarification @samuelAle :bow: I believe in order to be consitent, the desired behavior should be that the Framework does not set any default values for `identitySource` but rather checks if `resultTtlInSeconds` is non-zero and throws an error if `identitySource` is not defined - as mentioned, it will be consistent with the way it currently works for `httpApi` authorizers and I believe as it should be from beginning. 

So, the potential implementation proposal in my opinion should be like this:
1. Introduce a deprecation in current major version (2.x) that from new major, the `identitySource` won't be set by default and if they want to keep the previous behavior, they need to set it explicitly to `method.request.header.Authorization`. The deprecation should not be emitted if user already explicitly specifies the `indentitySource`. 
2. With the next major, we should drop setting the default and remove the deprecation, while also introducing the logic that will throw an error if `resultTtlInSeconds` is non-zero and `identitySource` is missing. 

What do you think @samuelAle ?

**Comment by samuelAle** [comment]
Apologies for the late reply. I think this strategy you mention is practical. Let's move forward with that.

**Comment by pgrzesik** [comment]
Hello everyone interested in this issue :wave: After initial implementation, I realised that I didn't understand the full implications very well and I've changed the approach a little bit. With #9825, the deprecation will be emitted for authorizers of the `request` type that have caching set to `0` and do not have explicit `identitySource` specified and in next major (v3), the default will be no longer set. At the moment, to hide deprecation in such situations you will be able to either specify it explicitly or set `identitySource` to `null` which will disable this setting. 

tldr for people that want to have this disabled: set `identitySource` to `null` explicitly to disable this setting

**Comment by SrushithR** [comment]
> set identitySource to null which will disable this setting.

Hey @pgrzesik, if I set the `identitySource` to `null`, I am getting the following error:

```
 An error occurred: 4NestedStack - [/Resources/AuthorizerLambdaApiGatewayAuthorizer/Type/IdentitySource] 'null' values are not allowed in templates.
```

**Comment by pgrzesik** [comment]
Hello @SrushithR :wave: What version of the Framework are you using?

**Comment by SrushithR** [comment]
Hi @pgrzesik, it is:
```
Framework Core: 2.57.0 (local)
Plugin: 5.4.4
SDK: 4.3.0
Components: 3.16.0
```

**Comment by pgrzesik** [comment]
Thanks for sharing @SrushithR :bow: Yes, that's the case, but we also have logic that strips all `null` properties from the template that should take care of it. Do you maybe have a small reproducible example that I could run on my side? 

I've run a simple example with `identitySource: null` and that's the CF for authorizer:

```
"NameApiGatewayAuthorizer": {
      "Type": "AWS::ApiGateway::Authorizer",
      "Properties": {
        "AuthorizerResultTtlInSeconds": 0,
        "Name": "Name",
        "RestApiId": {
          "Ref": "ApiGatewayRestApi"
        },
        "IdentityValidationExpression": "someRegex",
        "AuthorizerUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region"
              },
              ":lambda:path/2015-03-31/functions/",
              "arn:xxx:xxx:Lambda-Name",
              "/invocations"
            ]
          ]
        },
        "Type": "REQUEST"
      }
    }
```

**Comment by SrushithR** [comment]
Thanks for the quick response, Piotr! Here is my lambda configuration:
```
paymentsV2:
 handler: PaymentV2/index.payment
 events:
   - http:
       path: /v2/payments
       method: ANY
       cors: true
       authorizer:
         name: commonAuthorizerLambda
         resultTtlInSeconds: 0
         type: request
         identitySource: null
```
Can you share the lambda configuration with the authorizer set as `null` that you used to generate the sample CF

**Comment by pgrzesik** [comment]
Sure, that's how it looks like:

```
functions:
  exampleFunction:
    handler: handler.hello
    events:
      - http:
          path: posts/create
          method: post
          authorizer:
            arn: arn:xxx:xxx:Lambda-Name
            type: request
            identitySource: null
            resultTtlInSeconds: 0
            identityValidationExpression: someRegex
```

**Comment by SrushithR** [comment]
Yes, we use the following plugins:
```
plugins:
  - serverless-pseudo-parameters
  - serverless-plugin-split-stacks
  - serverless-dotenv-plugin
  - serverless-prune-plugin
  - serverless-offline-sqs
  - serverless-offline
  - serverless-offline-scheduler
  - serverless-offline-aws-eventbridge
  ```
Is there any way to figure out which plugin could be causing the issue?

**Comment by pgrzesik** [comment]
I would suspect that if any of them, it would be `serverless-plugin-split-stacks` as it deals with raw CloudFormation and it might interfere with the Framework's logic to strip `null`s

**Comment by SrushithR** [comment]
Yes, even I am suspecting the same. I will try to see if I can figure anything from the `serverless-plugin-split-stacks` plugin. Thanks for the quick responses @pgrzesik. Appreciate it! 🙌

**Comment by SrushithR** [comment]
Hey @pgrzesik & others who are also facing the same issue and looking for a solution. This is what I did:

Instead of using the `authorizer` configuration provided by default on serverless framework, I went ahead and created a customer resource as below:
```
resources:
    Resources:
        CustomAuthorizer:
            Type: AWS::ApiGateway::Authorizer
            Properties:
            AuthorizerResultTtlInSeconds: 0
            AuthorizerUri: !Join
                - ""
                - - "arn:"
                - !Ref "AWS::Partition"
                - ":apigateway:"
                - !Ref "AWS::Region"
                - ":lambda:path/2015-03-31/functions/"
                - !GetAtt
                    - AuthorizerLambdaFunction
                    - Arn
                - /invocations
            Name: authorizer
            RestApiId: !Ref ApiGatewayRestApi
            Type: REQUEST
```

And consume this authorizer in my lambda configuration:
```
functions:
  estimates:
    handler: EstimatesV2/index.estimates
    events:
      - http:
          path: /v1/estimates/{proxy+}
          method: ANY
          cors: ${file(./resource/CORSHeaders.yml)}
          authorizer:
            type: request
            authorizerId:
              Ref: CustomAuthorizer
```

This way, I was able to create the API Gateway Authorizer as a custom resource without the `identitySource` and consume it in the lambda function

**Comment by sidigi** [comment]
`identitySource: []`
